### PR TITLE
API cleanup with respect to contexts

### PR DESCRIPTION
**Pull request by real-or-random**
- Documents the changes made to contexts that came with the transition to only static tables
 - Rename the static context
 - Expose the self tests for users of the static context

These are user-visible changes, so this is also a testbed for the changelog. I refrained from linking to the PR as the template proposed for multiple reasons:
 - Including the PR title is not in line with the suggestions in https://keepachangelog.com/en/1.0.0/ What they do instead is really just describe what parts in the API have changed, and I tend to agree that this is better for the user. The PR title can be rather technical or saying not too much, and a PR can correspond to multiple changes (for example this PR).
 - I'd be open to Include a link to the PR but I feel it's a little bit annoying because you get the PR number only after creating the PR... Also I think the changelog is addressed to our users, not our devs. It should be just a log of user-visible changes (syntax and visible API behavior), so we anyway shouldn't include every PR. 

A normal strategy would be to create a release first, then make the API changes, but in this case, this would be somewhat confusing. The important thing here is the changes made to context flags and these are already for quite some time, so it makes sense to document these even for the initial release.

### CODE REVIEW

File: include/secp256k1.h
Comment (real-or-random) [inline include/secp256k1.h:199]: Unfortunately, it's not possible to apply a `deprecated` attribute to a macro. We could turn this into a `static const` but this potentially breaks existing code. Maybe something to consider for the next API version.

Comment (apoelstra) [reply include/secp256k1.h:199]: It would be easier for other language bindings if these were non-`static` `const`s ... as either macros or statics they aren't exported so we just have to reproduce the values in the target language.

**Comment by apoelstra** [comment]
Thanks for this! It's super helpful that all the recent context changes are now visible in the docs.

**Review by jonasnick — COMMENTED** [review]
Concept ACK

> I refrained from linking to the PR

That's reasonable. It should be possible to use `git blame` on the change log to find the exact code changes. Can you remove the change log template suggestion to add "Title with link to Pull Request"? Or just remove the template part entirely (starting with `## [MAJOR.MINOR.PATCH] - YYYY-MM-DD`)?

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h:231]: added a comment to address that

**Comment by real-or-random** [comment]
Now that you mention the tests, I realize that many of the tests still use a lot of variants of the contexts, though we could simply check if all (non-static) contexts are equal (using memcpy). Let me add this.

But what about all the test code?  What we currently have is not wrong, it's just not efficient because we may run the same tests multiple tests. Should I simplify it? If yes, I don't mind doing it here but it could also be done in a separate PR.

**Comment by real-or-random** [comment]
> @real-or-random Actually, in what way is the static context not usable for signing operations?

Oh, we could make it usable for signing. (At the moment the relevant parts of the object are simply not initialized but that will be easy to fix, in particular now that https://github.com/bitcoin-core/secp256k1/pull/1120 has been merged.)

The actual question is whether we would like to make it usable for signing. The static context cannot be randomized, so you'll never get protection by random blinding. I think this would not be unreasonable (and simplify the API and the docs a lot). But I'm not sure if we want to encourage this mode of usage. I tend to the opposite: Now that dynamic context creation is fast and doesn't need a lot of memory, we should encourage people to use that route.

With the current API, you could argue that anyway a lot of users are probably never adding randomness, so it won't make a lot of difference. But I think when evaluating this, we should keep in mind that a future context API will hopefully make it easier to randomize, e.g., context creation should directly accept a seed, and some of the (standard) signing API calls should rerandomize the context etc...  

(This was one of the questions in https://github.com/bitcoin-core/secp256k1/pull/988#issuecomment-938650194, and there I decided to not change the API because we could always do it later.)

**Comment by sipa** [comment]
@real-or-random Hard question. On the one hand, yes, I think we do want to encourage using properly-(re)randomized signing contexts to get maximal blinding protection. But on the other hand, I also don't think we should go as far as saying that one cannot use the signing API without forcing randomization; I could imagine platforms where randomness may actually be unavailable at signing time.

**Comment by real-or-random** [comment]
> On the one hand, yes, I think we do want to encourage using properly-(re)randomized signing contexts to get maximal blinding protection. But on the other hand, I also don't think we should go as far as saying that one cannot use the signing API without forcing randomization;

Okay, then the current API does exactly this. You are not forced to call `context_randomize` but you don't get a convenience bonus from the static context.

**Comment by sipa** [comment]
I feel it would be simpler to do make the static context usable for signing functions, and then remove the "initialized for signing" and "initialized for verification" conditions in the API documentation everywhere.

**Comment by jonasnick** [comment]
I don't think "initialized for signing" is very confusing (it would be clearer if the flag was explicitly mentioned). The flag is not deprecated after all. Also the static context explicitly mentions that it's not initialized for signing. However, if it's not too complicated to implement, sipa's approach (which involves deprecating the signing flag?) seems to be a significant simplification of the initial release API.

**Comment by sipa** [comment]
> The PR (even in its current form) already removes all "initialized for verification". Or are you saying I missed some?

I'm confused why I said that. I must have been looking at the wrong branch... Please disregard.

**Comment by real-or-random** [comment]
I added some WIP commits to demonstrate how another version would look like. This now tries to avoid the "historical" terminology entirely. As a consequence, this also introduces the new flag `SECP256K1_CONTEXT_DEFAULT` (name inspired by https://github.com/bitcoin-core/secp256k1/issues/559 and up to bikeshedding of course).

WIP because
 - [ ] the actual flags are not yet updated
 - [ ] this does not touch all occurrences of "initialized for signing"
 - [ ] and this will need some paragraph formatting. 
 - [ ] examples need to be updated

I can take care of all of this once know which version we want to have.

Please have a look and tell me if you prefer this new version (maybe with some changes) or instead allowing secret key ops with the static context. I still lean towards disallowing secret key ops with the static context because allowing it would create an incentive to skip randomization:  

>> On the one hand, yes, I think we do want to encourage using properly-(re)randomized signing contexts to get maximal blinding protection. But on the other hand, I also don't think we should go as far as saying that one cannot use the signing API without forcing randomization;

> Okay, then the current API does exactly this. You are not forced to call `context_randomize` but you don't get a convenience bonus from the static context.

**Review by jonasnick — COMMENTED** [review]
I think your WIP commits are an improvement over the previous version. Regarding the exact wording, if we'd add a new function that operates on secret keys and requires a new context flag, would we document it as

```
 *  Args:    ctx:       pointer to a context object, configured with flag SECP256K1_CONTEXT_FOO (not secp256k1_context_static).
```
?

**Comment by real-or-random** [comment]
> ```
>  *  Args:    ctx:       pointer to a context object, configured with flag SECP256K1_CONTEXT_FOO (not secp256k1_context_static).
> ```

Yeah, that's roughly what I had in mind. Maybe s/configured/created/

**Comment by real-or-random** [comment]
Note: This approach got Concept ACKed at the IRC meeting two weeks ago, so it's currently up to me to unWIP the PR.

**Comment by real-or-random** [comment]
Ready for review.

This probably could have been split into multiple PRs in hindsight but I think it's still manageable.

One open question (that can be resolved after this PR) is whether and how we want to tidy the tests... At the moment they still use deprecated flags like `SECP256K1_CONTEXT_NONE`.

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h:240]: fixed

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h]: fixed

**Review by jonasnick — COMMENTED** [review]
ACK mod nits

As part of the review I created a [branch](https://github.com/jonasnick/secp256k1/commits/202207-selftest-api-jn) that removes the deprecated flags from the tests & benchmarks. I can open a PR after this one is merged.

> Can you remove the change log template suggestion to add "Title with link to Pull Request"? Or just remove the template part entirely (starting with ## [MAJOR.MINOR.PATCH] - YYYY-MM-DD)?

I can do this when I rebase #1055.

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h:853]: The pragmatic why I added this paragraph is to make sure we're not conveying the wrong impression that ECDH is protected. Without this paragraph, you'd probably think that ECDH is protected (because secret keys are involved.) 

So this violates the module separation but I also think the context (and its docs)  is somehow global independent of the modules in some sense.  

> Or is this paragraph intended to calrify the "computations involving secret keys" after which it makes sense to rerandomize?

That's another purpose, yes.

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h]: fixed by rephrasing

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h]: fixed

**Review by jonasnick — COMMENTED** [review]
ACK 92e7293d6129e73ecd1226e6f30930ace4f569de

**Review by apoelstra — APPROVED** [review]
utACK 92e7293d6129e73ecd1226e6f30930ace4f569de

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h:641]: Strictly speaking yes. I omitted this because it's perhaps obvious, it adds clutter everywhere, and confuses the user more than it helps: cloning the static context is a pretty obscure operation. (Is there any valid use case even?) I think we could in fact disallow this operation. I doubt this hurts anyone, and it would arguably make the API docs even simpler. But I didn't add this to this PR because it would technically be a backward incompatible change, and its of limited use: We cannot prevent the user from simply copying the memory, e.g., using memcpy.   

Now that you bring this up, what we could do is to add a further comment to the docs of `_context_clone` or the docs of the static context? I think would be the best way to document that the restrictions Do you want me to do this?

Comment (real-or-random) [reply include/secp256k1.h:641]: > Just pointing out that the static context can't be cloned should be sufficient.

"pointing out that the static context can't be cloned" would mean forbidding it. Or do you mean "pointing out that it should not be cloned"?

**Comment by jonasnick** [comment]
@real-or-random I'd be in favor of calling the non-deprecated flag NONE and keep it being the neutral element wrt `|` as you're suggesting. When I wrote the code you reference I had to stop for a moment and asked myself the exact question that you mentioned.

I'm a bit confused about versioning. If the user has a new header and an old library, and is compelled to replace their `VERIFY` flags with `NONE` after reading the header, they will get the wrong context even when the `SIGN` bit is set if the library is old enough. The same is true already for the current version of the PR if the user replaces `VERIFY` with `DEFAULT`.

**Comment by jonasnick** [comment]
I'd be in favor of calling the non-deprecated flag NONE and keep it being the neutral element wrt `|` as you're suggesting. When I wrote the code you reference I had to stop for a moment and asked myself the exact question that you mentioned.

File: include/secp256k1.h
Comment (real-or-random) [reply include/secp256k1.h:639]: > I think for now it's better to just outlaw cloning the static context, as I think there is no point? 

Agreed. I'm in the middle of doing this but it's a small mess and it should not hold up this PR. I'll open a follow up PR on top of this..

**Review by jonasnick — COMMENTED** [review]
ACK 4386a2306c2b8cf9ad3040d8010e4295f6f01490

**Pull request merged**
